File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int n = nums.length;
        int l = 0, r =n-1;
        int st = -1;
        while(l<=r){
            int mid = l+(r-l)/2;
            if(nums[mid]>target) r = mid-1;
            else if(nums[mid]<target) l = mid+1;
            else{
                st = mid;
                if(mid>0 && nums[mid]==nums[mid-1]) r = mid-1;
                else break;
            }
        }
        int[] a = {-1, -1};
        if(st==-1) return a;
        l = st; r = n-1;
        int ed = -1;
        while(l<=r){
            int mid = l+(r-l)/2;
            if(nums[mid]>target) r = mid-1;
            else if(nums[mid]<target) l = mid+1;
            else{
                ed = mid;
                if(mid<n-1 && nums[mid]==nums[mid+1]) l = mid+1;
                else break;
            }
        }
        a[0] = st; a[1] = ed;
        return a;
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def find_first(nums, target):
            left, right = 0, len(nums) - 1
            first = -1
            while left <= right:
                mid = (left + right) // 2
                if nums[mid] == target:
                    first = mid
                    right = mid - 1  
                elif nums[mid] < target:
                    left = mid + 1
                else:
                    right = mid - 1
            return first
        def find_last(nums, target):
            left, right = 0, len(nums) - 1
            last = -1
            while left <= right:
                mid = (left + right) // 2
                if nums[mid] == target:
                    last = mid
                    left = mid + 1  
                elif nums[mid] < target:
                    left = mid + 1
                else:
                    right = mid - 1
            return last
        start = find_first(nums, target)
        end = find_last(nums, target)
        return [start, end]